2-0-data-structures-and-algorithms/src/test/java/com/bobocode/bst/BinarySearchTreeTest.java

```diff
@@ -1,82 +1,249 @@
+    }
+
+    @Test
+    @Order(4)
+    void shouldCreateTreeFromListOfElements() {
+        tree = RecursiveBinarySearchTree.of(someElements);
+        for (var e : someElements) {
+            assertThat(contains(getRootObject(), e)).isTrue();
         }
-        assertThat(bst.size()).isEqualTo(someElements.length);
+
+        assertThat(getInnerSize()).isEqualTo(someElements.length);
     }
 
+    @Test
+    @Order(5)
+    void shouldThrowNullPointerExceptionIfInsertNull() {
+        fillTestTree(someElements);
+        assertThatNullPointerException().isThrownBy(() -> tree.insert(null));
+    }
+
+    @Test
+    @Order(6)
+    void shouldInsertUniqueElements() {
+        tree = new RecursiveBinarySearchTree<>();
+        for (Integer e : someElements) {
+            assertThat(contains(getRootObject(), e)).isFalse(); //does not contain
+            assertThat(tree.insert(e)).isTrue(); //do insert
+            assertThat(contains(getRootObject(), e)).isTrue(); //and contains
+        }
+    }
+
+    @Test
+    @Order(7)
+    void shouldIncreaseSizeWhenInsertUnique() {
+        tree = new RecursiveBinarySearchTree<>();
+
+        tree.insert(10);
+        tree.insert(15);
+        tree.insert(20);
+
+        assertThat(getInnerSize()).isEqualTo(3);
+    }
+
+    @Test
+    @Order(8)
+    void shouldNotInsertNonUniqueElements() {
+        fillTestTree(someElements);
+
+        assertThat(tree.insert(10)).isFalse();
+        assertThat(tree.insert(11)).isFalse();
+        assertThat(tree.insert(12)).isFalse();
+    }
+
+    @Test
+    @Order(9)
+    void shouldNotIncreaseSizeWhenInsertNotUnique() {
+        fillTestTree(someElements);
+
+        assertThat(getInnerSize()).isEqualTo(someElements.length);
+
+        tree.insert(10);
+        tree.insert(11);
+        tree.insert(12);
+
+        assertThat(getInnerSize()).isEqualTo(someElements.length);
+    }
+
+    @Order(10)
     @ParameterizedTest
     @MethodSource("depthArguments")
-    @Order(4)
     void depth(Integer[] elements, int depth) {
-        BinarySearchTree<Integer> bst = RecursiveBinarySearchTree.of(elements);
-        assertThat(bst.depth()).isEqualTo(depth);
+        fillTestTree(elements);
+        assertThat(tree.depth()).isEqualTo(depth);
     }
 
     @Test
-    @Order(5)
+    @Order(11)
     void inorderTraversal() {
-        BinarySearchTree<Integer> bst = RecursiveBinarySearchTree.of(someElements);
+        fillTestTree(someElements);
         Integer[] sortedElements = Arrays.copyOf(someElements, someElements.length);
         Arrays.sort(sortedElements);
 
-        List<Integer> traversedElements = new ArrayList<>(bst.size());
-        bst.inOrderTraversal(traversedElements::add);
+        List<Integer> traversedElements = new ArrayList<>(getInnerSize());
+        tree.inOrderTraversal(traversedElements::add);
 
         assertThat(traversedElements).isEqualTo(List.of(sortedElements));
     }
 
+    @Test
+    @Order(12)
+    void shouldInsertToRootIfEmpty() {
```

Comment:
@erneman1 if you test `insert` method, please move these tests up and change the order